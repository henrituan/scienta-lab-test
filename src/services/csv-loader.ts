import fs from 'fs';
import path from 'path';
import { parse } from 'csv-parse';

export const loadCSV = <T>(filename: string): Promise<T[]> => {
  return new Promise((resolve, reject) => {
    const filePath = path.join(process.cwd(), 'public/data', filename);
    const fileContent = fs.readFileSync(filePath, { encoding: 'utf-8' });

    parse(
      fileContent,
      {
        columns: true,
        skip_empty_lines: true,
      },
      (err, records) => {
        if (err) {
          reject(err);
        } else {
          resolve(records);
        }
      },
    );
  });
};
