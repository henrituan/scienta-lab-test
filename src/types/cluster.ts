export type ClusterRawData = {
  cluster_id: string;
  '%female': string;
  avg_age: string;
  symptoms: string;
};

export type Cluster = {
  clusterId: number;
  avgAge: number;
  femalePercent: number;
  malePercent: number;
  symptoms: string[];
};
