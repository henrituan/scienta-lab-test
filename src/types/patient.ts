export type PatientRawData = {
  patient_id: string;
  cluster_id: string;
  coordinates: string;
};

export type Patient = {
  patientId: number;
  clusterId: number;
  coordinates: {
    x: number;
    y: number;
  };
};
