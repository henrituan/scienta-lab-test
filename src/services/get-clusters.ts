import { Cluster, ClusterRawData } from '@/types/cluster';

import { loadCSV } from './csv-loader';

const getSymptoms = (rawStr: string): string[] => {
  // Remove outer parentheses
  const content = rawStr.slice(1, -1);

  // Match 1 item between quotes
  const regex = /'([^']*)'(?:,\s*|\s*$)/g;
  const items = [];
  let match;

  while ((match = regex.exec(content)) !== null) {
    items.push(match[1]);
  }

  return items;
};

export const getClusters = async (): Promise<Cluster[]> => {
  const records = await loadCSV<ClusterRawData>('Clusters.csv');

  return records.map((record) => {
    const femalePercent = parseFloat(record['%female']);

    return {
      clusterId: parseInt(record['cluster_id']),
      avgAge: parseFloat(record['avg_age']),
      femalePercent,
      malePercent: 100 - femalePercent,
      symptoms: getSymptoms(record['symptoms']),
    };
  });
};
