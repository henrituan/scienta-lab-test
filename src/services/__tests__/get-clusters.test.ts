import { vi, describe, it, expect } from 'vitest';

import { Cluster, ClusterRawData } from '@/types/cluster';

import { getClusters } from '../get-clusters';
import { loadCSV } from '../csv-loader';

const mockedCluster: ClusterRawData[] = [
  {
    cluster_id: '4',
    '%female': '21',
    avg_age: '74.73954504286607',
    symptoms:
      "('Antibiotics', 'Hair loss on other parts of the body', 'Small, raised bumps', 'Premature whitening or graying of hair', 'Minoxidil', 'Psoriasis', 'Swollen and stiff joints', 'Depigmentation (for widespread vitiligo)', 'Scalp sensitivity or itching', 'Dry, cracked skin that may bleed')",
  },
  {
    cluster_id: '5',
    '%female': '82',
    avg_age: '74.18689944450973',
    symptoms:
      "('Leaking, foul-smelling abscesses', 'JAK inhibitors', 'Platelet-rich plasma therapy', 'Vitamin D analogues', 'Atopic Dermatitis (Eczema)', 'Biologics (e.g., TNF-alpha inhibitors)', 'Patchy loss of skin color', 'Hormone therapy', 'Thickened, pitted  or ridged nails', 'Swollen and stiff joints')",
  },
  {
    cluster_id: '6',
    '%female': '57',
    avg_age: '42.34856316329995',
    symptoms:
      "('Skin grafting', 'Phototherapy', 'Small, raised bumps', 'Vitamin D analogues', 'Thickened, pitted or ridged nails', 'Raw, sensitive, swollen skin from scratching', 'Red, scaly patches on skin', 'Hair loss on other parts of the body', 'Topical corticosteroids', 'Atopic Dermatitis (Eczema)')",
  },
  {
    cluster_id: '7',
    '%female': '88',
    avg_age: '36.19179340285662',
    symptoms:
      "('Dupilumab (biologic)', 'Retinoids', 'Skin grafting', 'Topical calcineurin inhibitors', 'JAK inhibitors', 'Patchy loss of skin color', 'Scalp sensitivity or itching', 'Atopic Dermatitis (Eczema)', 'Premature whitening or graying of hair', 'Small, raised bumps')",
  },
];

const mocks = vi.hoisted(() => {
  return {
    loadCSV: vi.fn(),
  };
});

vi.mock('../csv-loader', () => {
  return {
    loadCSV: mocks.loadCSV,
  };
});

vi.mocked(loadCSV).mockResolvedValue(mockedCluster);

describe('get-clusters', async () => {
  const results = await getClusters();

  it('should return correct format', async () => {
    expect(results[0]).toEqual({
      clusterId: 4,
      avgAge: 74.73954504286607,
      femalePercent: 21,
      malePercent: 79,
      symptoms: [
        'Antibiotics',
        'Hair loss on other parts of the body',
        'Small, raised bumps',
        'Premature whitening or graying of hair',
        'Minoxidil',
        'Psoriasis',
        'Swollen and stiff joints',
        'Depigmentation (for widespread vitiligo)',
        'Scalp sensitivity or itching',
        'Dry, cracked skin that may bleed',
      ],
    } satisfies Cluster);
  });

  it('should match snapshot', async () => {
    expect(results).toMatchSnapshot();
  });
});
