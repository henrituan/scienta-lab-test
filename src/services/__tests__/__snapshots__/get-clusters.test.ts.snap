// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`get-clusters > should match snapshot 1`] = `
[
  {
    "avgAge": 74.73954504286607,
    "clusterId": 4,
    "femalePercent": 21,
    "malePercent": 79,
    "symptoms": [
      "Antibiotics",
      "Hair loss on other parts of the body",
      "Small, raised bumps",
      "Premature whitening or graying of hair",
      "Minoxidil",
      "Psoriasis",
      "Swollen and stiff joints",
      "Depigmentation (for widespread vitiligo)",
      "Scalp sensitivity or itching",
      "Dry, cracked skin that may bleed",
    ],
  },
  {
    "avgAge": 74.18689944450973,
    "clusterId": 5,
    "femalePercent": 82,
    "malePercent": 18,
    "symptoms": [
      "Leaking, foul-smelling abscesses",
      "JAK inhibitors",
      "Platelet-rich plasma therapy",
      "Vitamin D analogues",
      "Atopic Dermatitis (Eczema)",
      "Biologics (e.g., TNF-alpha inhibitors)",
      "Patchy loss of skin color",
      "Hormone therapy",
      "Thickened, pitted  or ridged nails",
      "Swollen and stiff joints",
    ],
  },
  {
    "avgAge": 42.34856316329995,
    "clusterId": 6,
    "femalePercent": 57,
    "malePercent": 43,
    "symptoms": [
      "Skin grafting",
      "Phototherapy",
      "Small, raised bumps",
      "Vitamin D analogues",
      "Thickened, pitted or ridged nails",
      "Raw, sensitive, swollen skin from scratching",
      "Red, scaly patches on skin",
      "Hair loss on other parts of the body",
      "Topical corticosteroids",
      "Atopic Dermatitis (Eczema)",
    ],
  },
  {
    "avgAge": 36.19179340285662,
    "clusterId": 7,
    "femalePercent": 88,
    "malePercent": 12,
    "symptoms": [
      "Dupilumab (biologic)",
      "Retinoids",
      "Skin grafting",
      "Topical calcineurin inhibitors",
      "JAK inhibitors",
      "Patchy loss of skin color",
      "Scalp sensitivity or itching",
      "Atopic Dermatitis (Eczema)",
      "Premature whitening or graying of hair",
      "Small, raised bumps",
    ],
  },
]
`;
